File: swift/Tests/RecommendAITests/ClientTests.swift
import Foundation
import XCTest
@testable import RecommendAI

// MARK: - Mock URLProtocol

final class MockURLProtocol: URLProtocol {
    // Map of path -> (statusCode, responseData)
    static var handlers: [(check: (URLRequest) -> Bool, response: (Int, Data))] = []

    static func stub(path: String, status: Int, json: Any) {
        let data = try! JSONSerialization.data(withJSONObject: json)
        handlers.append((
            check: { $0.url?.path.hasPrefix(path) ?? false },
            response: (status, data)
        ))
    }

    static func reset() { handlers = [] }

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        let match = MockURLProtocol.handlers.first { $0.check(request) }
        if let (status, data) = match?.response {
            let response = HTTPURLResponse(
                url: request.url!,
                statusCode: status,
                httpVersion: nil,
                headerFields: ["Content-Type": "application/json"]
            )!
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
        } else {
            client?.urlProtocol(self, didFailWithError: URLError(.fileDoesNotExist))
        }
        client?.urlProtocolDidFinishLoading(self)
    }

    override func stopLoading() {}
}

// MARK: - Helpers

func makeClient() -> RecommendAIClient {
    let cfg = URLSessionConfiguration.ephemeral
    cfg.protocolClasses = [MockURLProtocol.self]
    let session = URLSession(configuration: cfg)
    return RecommendAIClient(
        apiKey: "test-key",
        config: ClientConfig(baseURL: "http://localhost:8080"),
        session: session
    )
}

// MARK: - Tests

final class RecommendAIClientTests: XCTestCase {

    override func setUp() {
        super.setUp()
        MockURLProtocol.reset()
    }

    // ── ping ─────────────────────────────────────────────────────────────────

    func testPingReturnsTrue() async {
        MockURLProtocol.stub(path: "/health", status: 200, json: [:])
        let result = await makeClient().ping()
        XCTAssertTrue(result)
    }

    func testPingReturnsFalse() async {
        MockURLProtocol.stub(path: "/health", status: 503, json: [:])
        let result = await makeClient().ping()
        XCTAssertFalse(result)
    }

    // ── recommendations ───────────────────────────────────────────────────────

    func testRecommendationsSimilarCallsCorrectPath() async throws {
        MockURLProtocol.stub(path: "/api/recommendations/similar/item99", status: 200, json: [
            "recommendations": [
                ["item_id": "item2", "score": 0.8, "reason": "", "metadata": [:] as [String: Any]]
            ]
        ])
        let client = makeClient()
        let recs = try await client.recommendations.similar(itemId: "item99", limit: 10)
        XCTAssertEqual(recs.count, 1)
        XCTAssertEqual(recs[0].itemId, "item2")
    }

    func testRecommendationsPopularPassesCategoryParam() async throws {
        MockURLProtocol.stub(path: "/api/recommendations/popular", status: 200, json: [
            "recommendations": [
                ["item_id": "book1", "score": 0.7, "reason": "", "metadata": [:] as [String: Any]]
            ]
        ])
        let client = makeClient()
        let recs = try await client.recommendations.popular(limit: 5, category: "books")
        XCTAssertEqual(recs.count, 1)
        XCTAssertEqual(recs[0].itemId, "book1")
    }

    // ── items ─────────────────────────────────────────────────────────────────

    func testItemsUpsertPostsToBulkEndpoint() async throws {
        MockURLProtocol.stub(path: "/api/items/bulk", status: 200, json: [
            "items": [
                ["item_id": "itemA", "properties": [:] as [String: Any]]
            ]
        ])
        let client = makeClient()
        let items = try await client.items.upsert(items: [["item_id": "itemA", "properties": ["name": "Book A"]]])
        XCTAssertEqual(items.count, 1)
        XCTAssertEqual(items[0].itemId, "itemA")
    }

    // ── error handling ────────────────────────────────────────────────────────

    func testAuthenticationErrorOn401() async {
        MockURLProtocol.stub(path: "/api/recommendations/similar/x", status: 401, json: ["detail": "invalid api key"])
        let client = makeClient()
        do {
            _ = try await client.recommendations.similar(itemId: "x", limit: 5)
            XCTFail("Expected error")
        } catch RecommendAIError.authentication {
            // pass
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testNotFoundErrorOn404() async {
        MockURLProtocol.stub(path: "/api/recommendations/similar/x", status: 404, json: ["detail": "not found"])
        let client = makeClient()
        do {
            _ = try await client.recommendations.similar(itemId: "x", limit: 5)
            XCTFail("Expected error")
        } catch RecommendAIError.notFound {
            // pass
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testRateLimitErrorOn429() async {
        MockURLProtocol.stub(path: "/api/recommendations/similar/x", status: 429, json: ["detail": "rate limit"])
        let client = makeClient()
        do {
            _ = try await client.recommendations.similar(itemId: "x", limit: 5)
            XCTFail("Expected error")
        } catch RecommendAIError.rateLimit {
            // pass
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }
}

File: rust/tests/client_tests.rs
use mockito::Server;
use recommendai::{ClientConfig, RecommendAIClient};
use serde_json::json;

fn make_client(server: &Server) -> RecommendAIClient {
    RecommendAIClient::with_config(
        "test-key",
        ClientConfig {
            base_url: server.url(),
            timeout: std::time::Duration::from_secs(5),
        },
    )
    .expect("client creation should succeed")
}

// ── ping ──────────────────────────────────────────────────────────────────────

#[test]
fn ping_returns_true_on_200() {
    let mut server = Server::new();
    let _m = server.mock("GET", "/health").with_status(200).create();

    let client = make_client(&server);
    assert!(client.ping());
}

#[test]
fn ping_returns_false_on_503() {
    let mut server = Server::new();
    let _m = server.mock("GET", "/health").with_status(503).create();

    let client = make_client(&server);
    assert!(!client.ping());
}

// ── recommendations ───────────────────────────────────────────────────────────

#[test]
fn recommendations_similar_calls_correct_path() {
    let mut server = Server::new();
    let body = json!({
        "recommendations": [
            { "item_id": "item2", "score": 0.8, "reason": "", "metadata": {} }
        ]
    });
    let _m = server
        .mock("GET", "/api/recommendations/similar/item99")
        .match_query(mockito::Matcher::UrlEncoded("limit".into(), "10".into()))
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let recs = client.recommendations().similar("item99", 10).unwrap();
    assert_eq!(recs.len(), 1);
    assert_eq!(recs[0].item_id, "item2");
}

#[test]
fn recommendations_popular_passes_category_param() {
    let mut server = Server::new();
    let body = json!({
        "recommendations": [
            { "item_id": "book1", "score": 0.7, "reason": "", "metadata": {} }
        ]
    });
    let _m = server
        .mock("GET", "/api/recommendations/popular")
        .match_query(mockito::Matcher::AllOf(vec![
            mockito::Matcher::UrlEncoded("limit".into(), "5".into()),
            mockito::Matcher::UrlEncoded("category".into(), "books".into()),
        ]))
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(body.to_string())
        .create();

    let client = make_client(&server);
    let recs = client
        .recommendations()
        .popular(5, Some("books"))
        .unwrap();
    assert_eq!(recs.len(), 1);
    assert_eq!(recs[0].item_id, "book1");
}

// ── items ─────────────────────────────────────────────────────────────────────

#[test]
fn items_upsert_posts_to_bulk_endpoint() {
    let mut server = Server::new();
    let resp_body = json!({
        "items": [
            { "item_id": "itemA", "properties": {}, "created_at": null, "updated_at": null }
        ]
    });
    let _m = server
        .mock("POST", "/api/items/bulk")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(resp_body.to_string())
        .create();

    let client = make_client(&server);
    let items = client
        .items()
        .upsert(vec![serde_json::from_value(json!({
            "item_id": "itemA",
            "properties": { "name": "Book A" }
        }))
        .unwrap()])
        .unwrap();
    assert_eq!(items.len(), 1);
    assert_eq!(items[0].item_id, "itemA");
}

// ── error handling ────────────────────────────────────────────────────────────

#[test]
fn authentication_error_on_401() {
    let mut server = Server::new();
    let _m = server
        .mock("GET", "/api/recommendations/similar/x")
        .match_query(mockito::Matcher::Any)
        .with_status(401)
        .with_header("content-type", "application/json")
        .with_body(json!({"detail": "invalid api key"}).to_string())
        .create();

    let client = make_client(&server);
    let err = client.recommendations().similar("x", 5).unwrap_err();
    assert!(matches!(err, recommendai::Error::Authentication(_)));
}

#[test]
fn not_found_error_on_404() {
    let mut server = Server::new();
    let _m = server
        .mock("GET", "/api/recommendations/similar/x")
        .match_query(mockito::Matcher::Any)
        .with_status(404)
        .with_header("content-type", "application/json")
        .with_body(json!({"detail": "not found"}).to_string())
        .create();

    let client = make_client(&server);
    let err = client.recommendations().similar("x", 5).unwrap_err();
    assert!(matches!(err, recommendai::Error::NotFound(_)));
}

#[test]
fn rate_limit_error_on_429() {
    let mut server = Server::new();
    let _m = server
        .mock("GET", "/api/recommendations/similar/x")
        .match_query(mockito::Matcher::Any)
        .with_status(429)
        .with_header("content-type", "application/json")
        .with_body(json!({"detail": "rate limit"}).to_string())
        .create();

    let client = make_client(&server);
    let err = client.recommendations().similar("x", 5).unwrap_err();
    assert!(matches!(err, recommendai::Error::RateLimit(_)));
}
